zio-test's provideLayer is broken

I have a test suite that uses `provideLayer`. Since 5f99983b, the tests fail without providing any output:

```
[info] No tests were executed
[error] Error during tests:
[error]         scalapb.zio_grpc.TestServiceSpec
````

I looked into this and I actually don't think the issue is related to that PR or even `ZLayer`. I was able to minimize to the following:

```scala
package zio.test

import zio.ZIO

object InitializerSpec extends ZIOBaseSpec {
  import zio.NeedsEnv.needsEnv

  val env = ZIO.unit

  def spec =
    suite("InitializerSpec") {
      test("some test")(assertCompletes)
    }.provideM(env)
}

// java.lang.ExceptionInInitializerError
// Caused by: java.lang.NullPointerException
```

I think the problem is actually related to the way we changed `DefaultRunnableSpec` to be a trait and early initialization issues in Scala. If we define a local value within the spec like `env` above it is fine to use in evaluating tests in the spec itself, but it looks like the members of that value are not yet initialized when we call `provide` or its variants on the spec. If we change `val env` to `lazy val env` that solves the problem. Also if we just inline it `provideM(ZIO.unit)` that is fine. And it is also fine to define a separate object like we used to in RC17 and define the value there.

So obviously there are a bunch of easy immediate solutions mentioned above but it is not great for users to have to think about this stuff. To fix we can make all of these methods on spec lazy. For example:

```scala
// Before
final def provideM[E1 >: E](zio: ZIO[Any, E1, R])(implicit ev: NeedsEnv[R]): Spec[Any, E1, T] =
  provideManaged(zio.toManaged_)

// After
final def provideM[E1 >: E](zio: => ZIO[Any, E1, R])(implicit ev: NeedsEnv[R]): Spec[Any, E1, T] =
  provideManaged(ZManaged.suspend(zio.toManaged_))
```

Unfortunately we have to think about this for a lot of the methods on Spec. Possibly we could write a `Spec.suspend` constructor that we could use to wrap a bunch of these but unfortunately I am not seeing a way to more systematically address this where no one ever has to think about this again in the future.

--- a/test-tests/shared/src/test/scala/zio/test/SpecSpec.scala
+++ b/test-tests/shared/src/test/scala/zio/test/SpecSpec.scala
@@ -3,9 +3,17 @@ package zio.test
 import zio.test.Assertion.{ equalTo, isFalse, isTrue }
 import zio.test.TestAspect.ifEnvSet
 import zio.test.TestUtils._
-import zio.{ Ref, UIO, ZIO, ZManaged }
+import zio.{ Has, Ref, UIO, ZIO, ZLayer, ZManaged }
 
 object SpecSpec extends ZIOBaseSpec {
+
+  type Module = Has[Module.Service]
+
+  object Module {
+    trait Service
+  }
+
+  val layer = ZLayer.succeed(new Module.Service {})
 
   def spec = suite("SpecSpec")(
     suite("provideManagedShared")(
@@ -76,6 +84,13 @@ object SpecSpec extends ZIOBaseSpec {
           passed <- isSuccess(mixedSpec.only(rootSuite))
         } yield assert(passed)(isFalse)
       }
+    ),
+    suite("provideLayer")(
+      testM("does not have early initialization issues") {
+        for {
+          _ <- ZIO.environment[Module]
+        } yield assertCompletes
+      }.provideLayer(layer)
     )
   )
 

--- a/test/shared/src/main/scala/zio/test/RunnableSpec.scala
+++ b/test/shared/src/main/scala/zio/test/RunnableSpec.scala
@@ -27,12 +27,13 @@ trait RunnableSpec[R, E] extends AbstractRunnableSpec {
   override type Environment = R
   override type Failure     = E
 
-  private val runSpec: URIO[TestLogger with Clock, Int] = for {
-    results     <- run
-    hasFailures <- results.exists { case TestCase(_, test, _) => test.map(_.isLeft); case _ => UIO.succeedNow(false) }
-    summary     <- SummaryBuilder.buildSummary(results)
-    _           <- TestLogger.logLine(summary.summary)
-  } yield if (hasFailures) 1 else 0
+  private def runSpec: URIO[TestLogger with Clock, Int] =
+    for {
+      results     <- run
+      hasFailures <- results.exists { case TestCase(_, test, _) => test.map(_.isLeft); case _ => UIO.succeedNow(false) }
+      summary     <- SummaryBuilder.buildSummary(results)
+      _           <- TestLogger.logLine(summary.summary)
+    } yield if (hasFailures) 1 else 0
 
   /**
    * A simple main function that can be used to run the spec.

